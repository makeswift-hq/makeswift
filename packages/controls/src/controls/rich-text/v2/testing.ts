import { z } from 'zod'

import { type SerializedRecord } from '../../../serialization'
import { type DataType } from '../../associated-types'
import { serialize } from '../../definition'
import {
  ControlInstance,
  DefaultControlInstance,
  type SendMessage,
} from '../../instance'

import { RichTextPluginControl } from './plugin'
import { RichTextDefinition } from './rich-text'

export const renderedNode = Symbol('rendered node')

export type RenderedNode = typeof renderedNode

type UserConfig = z.infer<typeof Definition.schema.userConfig>

class Definition extends RichTextDefinition<RenderedNode> {
  createInstance(sendMessage: SendMessage<any>): ControlInstance<any> {
    return new DefaultControlInstance(sendMessage)
  }

  serialize(): [SerializedRecord, Transferable[]] {
    return serialize(this.config, {
      type: Definition.type,
    })
  }

  toText(_data: DataType<RichTextDefinition<string>>): string {
    return ''
  }

  get pluginControls(): RichTextPluginControl[] {
    return []
  }

  pluginControlAt(_index: number) {
    return undefined
  }
}

export { Definition as RichTextDefinition }

export const RichText = (config?: UserConfig) => new Definition(config ?? {})

RichText.Mode = Definition.Mode
