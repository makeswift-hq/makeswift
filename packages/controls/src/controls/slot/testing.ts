import { SlotDefinition } from './slot'

export const renderedNode = Symbol('rendered node')

export type RenderedNode = typeof renderedNode

class Definition extends SlotDefinition<RenderedNode> {}

export function Slot(): Definition {
  return new Definition()
}

export { Definition as SlotDefinition }
