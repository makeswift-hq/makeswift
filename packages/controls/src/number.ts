import { ControlDataTypeKey } from "./common/data-type-key"

export const NumberControlDataTypeKey = ControlDataTypeKey

export const NumberControlDataTypeValueV1 = 'number::v1'

export type NumberControlDataV0 = number

export type NumberControlDataV1 = {
  [NumberControlDataTypeKey]: typeof NumberControlDataTypeValueV1
  value: number,
}

export type NumberControlData = NumberControlDataV0 | NumberControlDataV1

export const NumberControlType = 'makeswift::controls::number'

type NumberControlConfig = {
  label?: string
  labelOrientation?: 'vertical' | 'horizontal'
  defaultValue?: number
  min?: number
  max?: number
  step?: number
  suffix?: string
}

export type NumberControlDefinition<C extends NumberControlConfig = NumberControlConfig> = {
  type: typeof NumberControlType
  config: C
  version?: 1
}

export function Number<C extends NumberControlConfig>(
  config: C = {} as C,
): NumberControlDefinition<C> {
  return { type: NumberControlType, config, version: 1 }
}
