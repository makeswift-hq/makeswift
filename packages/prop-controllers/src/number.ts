import { z } from 'zod'
import { ControlDataTypeKey, Options, Types } from './prop-controllers'
import { P, match } from 'ts-pattern'

const numberPropControllerDataV0Schema = z.number()

type NumberPropControllerDataV0 = z.infer<
  typeof numberPropControllerDataV0Schema
>

const NumberPropControllerDataV1Type = 'prop-controllers::number::v1'

const numberPropControllerDataV1Schema = z.object({
  [ControlDataTypeKey]: z.literal(NumberPropControllerDataV1Type),
  value: z.number(),
})

type NumberPropControllerDataV1 = z.infer<
  typeof numberPropControllerDataV1Schema
>

export const numberPropControllerDataSchema = z.union([
  numberPropControllerDataV0Schema,
  numberPropControllerDataV1Schema,
])

export type NumberPropControllerData = z.infer<
  typeof numberPropControllerDataSchema
>

export type NumberOptions = Options<{
  preset?: NumberPropControllerData
  label?: string
  defaultValue?: number
  min?: number
  max?: number
  step?: number
  suffix?: string
  hidden?: boolean
}>

type NumberDescriptorV0<_T = NumberPropControllerDataV0> = {
  type: typeof Types.Number
  options: NumberOptions
}

type NumberDescriptorV1<_T = NumberPropControllerDataV1> = {
  type: typeof Types.Number
  version: 1
  options: NumberOptions
}

export type NumberDescriptor<_T = NumberPropControllerData> =
  | NumberDescriptorV0
  | NumberDescriptorV1

export type ResolveNumberPropControllerValue<T extends NumberDescriptor> =
  T extends NumberDescriptor ? number | undefined : never

/**
 * @deprecated Imports from `@makeswift/runtime/prop-controllers` are deprecated. Use
 * `@makeswift/runtime/controls` instead.
 */
export function Number(options: NumberOptions = {}): NumberDescriptorV1 {
  return { type: Types.Number, version: 1, options }
}

export function getNumberPropControllerDataNumber(
  data: NumberPropControllerData,
): number {
  return match(data)
    .with(
      { [ControlDataTypeKey]: NumberPropControllerDataV1Type },
      (v1) => v1.value,
    )
    .otherwise((v0) => v0)
}

export function createNumberPropControllerDataFromNumber(
  value: number,
  definition?: NumberDescriptor,
): NumberPropControllerData {
  return match(definition)
    .with(
      { version: 1 },
      P.nullish,
      () =>
        ({
          [ControlDataTypeKey]: NumberPropControllerDataV1Type,
          value,
        } as const),
    )
    .otherwise(() => value)
}
