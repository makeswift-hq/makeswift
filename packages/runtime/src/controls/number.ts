export type NumberControlData = number

export const NumberControlType = 'makeswift::controls::number'

type NumberControlConfig = {
  label?: string
  labelOrientation?: 'vertical' | 'horizontal'
  defaultValue?: number
  min?: number
  max?: number
  step?: number
  suffix?: string
}

export type NumberControlDefinition<C extends NumberControlConfig = NumberControlConfig> = {
  type: typeof NumberControlType
  config: C
}

export function Number<C extends NumberControlConfig>(
  config: C = {} as C,
): NumberControlDefinition<C> {
  return { type: NumberControlType, config }
}
