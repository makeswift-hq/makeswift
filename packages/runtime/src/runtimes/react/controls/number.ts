import { match } from 'ts-pattern'
import { NumberControlData, NumberControlDataTypeKey, NumberControlDataTypeValueV1, NumberControlDefinition } from '../../../controls'

export type NumberControlValue<T extends NumberControlDefinition> =
  undefined extends T['config']['defaultValue'] ? number | undefined : number

export function useNumber<T extends NumberControlDefinition>(
  data: NumberControlData | undefined,
  definition: T,
): NumberControlValue<T> {
  const value: number | undefined = match(data)
    .with({ [NumberControlDataTypeKey]: NumberControlDataTypeValueV1 }, (val) => val.value)
    .otherwise(val => val) ?? definition.config.defaultValue

  return value as NumberControlValue<T>
}
