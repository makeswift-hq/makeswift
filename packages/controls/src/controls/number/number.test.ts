import { testDefinition } from '../../testing/test-definition'

import { Number, NumberDefinition } from './number'

describe('Number', () => {
  describe('constructor', () => {
    test.each([1, 2, undefined])(
      'call with default value `%s` returns versioned definition',
      (value) => {
        expect(
          Number({
            label: 'Number',
            defaultValue: value,
          }),
        ).toMatchSnapshot()
      },
    )

    test('disallows extraneous properties', () => {
      Number({
        label: undefined,
        defaultValue: undefined,
        // @ts-expect-error
        extra: 'extra',
      })
    })
  })

  describe('assignability', () => {
    function assignTest(_def: NumberDefinition) {}
    assignTest(Number())
    assignTest(Number({ label: 'number' }))
    assignTest(Number({ defaultValue: 17 }))
    assignTest(Number({ label: 'number', defaultValue: -1 }))
    assignTest(
      Number({ defaultValue: -1 as number, labelOrientation: 'horizontal' }),
    )

    assignTest(Number({ defaultValue: undefined, min: 0, max: 100 }))
    assignTest(Number({ label: undefined, defaultValue: undefined }))
  })

  const invalidValues = [null, false, 'text']

  testDefinition(
    Number({ defaultValue: 17, label: 'visible' }),
    [21, 20],
    invalidValues,
  )

  testDefinition(
    Number({ label: 'Number' }),
    [50, 51, undefined],
    invalidValues,
  )
})
