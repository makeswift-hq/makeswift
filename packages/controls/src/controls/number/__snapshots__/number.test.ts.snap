// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Number constructor call with default value \`1\` returns versioned definition 1`] = `
NumberDefinition {
  "config": {
    "defaultValue": 1,
    "label": "Number",
  },
  "version": 1,
}
`;

exports[`Number constructor call with default value \`2\` returns versioned definition 1`] = `
NumberDefinition {
  "config": {
    "defaultValue": 2,
    "label": "Number",
  },
  "version": 1,
}
`;

exports[`Number constructor call with default value \`undefined\` returns versioned definition 1`] = `
NumberDefinition {
  "config": {
    "defaultValue": undefined,
    "label": "Number",
  },
  "version": 1,
}
`;

exports[`Number definition w/ config {"defaultValue":17,"label":"visible"} toData returns v1 data for \`20\` when definition version is 1 1`] = `
{
  "@@makeswift/type": "number::v1",
  "value": 20,
}
`;

exports[`Number definition w/ config {"defaultValue":17,"label":"visible"} toData returns v1 data for \`21\` when definition version is 1 1`] = `
{
  "@@makeswift/type": "number::v1",
  "value": 21,
}
`;

exports[`Number definition w/ config {"label":"Number"} toData returns v1 data for \`50\` when definition version is 1 1`] = `
{
  "@@makeswift/type": "number::v1",
  "value": 50,
}
`;

exports[`Number definition w/ config {"label":"Number"} toData returns v1 data for \`51\` when definition version is 1 1`] = `
{
  "@@makeswift/type": "number::v1",
  "value": 51,
}
`;

exports[`Number definition w/ config {"label":"Number"} toData returns v1 data for \`undefined\` when definition version is 1 1`] = `undefined`;
